### ISSUE TITLE
Configuration of default log file

### ISSUE THREAD

**Issue by dokterbob** [sampsyo/beets#431]
The idea is to add a configuration option providing for a default value for the `-l` command-line parameter while maintaining legacy behaviour through command-line overrides.

This is an attempt to split #327 up into several byte-sized features which can be concretely implemented.

**Comment by brunal** [comment]
Currently most plugins use the "beets" logger. What's the desired level of diversity?
(a) every module has a different logger, called `__name__`
(b) single logger for the `beets` package, one logger per plugin (1) called "beets.plugin.$plugin_name" (2) called "beetsplug.$plugin_name"

**Comment by sampsyo** [comment]
Thanks for looking into this, @burnal. I'd like to do (b), and I'd lean toward (1) over (2) but it doesn't really matter too much.

And in case I haven't typed it nearby enough: I'd love for this to happen automatically (i.e., as part of the `BeetsPlugin` class) rather than having each plugin do it manually as a convention.

**Comment by brunal** [comment]
##### `_log` as a BeetsPlugin attribute
This causes one problem: many plugin provide top-level functions that use logging, e.g. `beetsplug.lyrics.fetch_url`. All those functions would have to turn into methods, or there would be 2 references to the logger: one is the attribute, the other the result of a manual `log = logging.getLogger('beets.' + __name__.split('.')[-1])`

##### Automatic logging level
See #430 and #432, issues related to the topic. The `logLevel` would be set to `WARNING` (a.k.a the default level) before running one of `import_stages` and `INFO` before running the `func` of one of `commands`. I still have to iron out the details but it seems OK.

**Comment by sampsyo** [comment]
Good points all. In general, either passing around the logger or moving functions to methods both seem like good ideas. Different strategies may be more appropriate for different situations, with the "duplicate logger" thing as an acceptable fallback when the other changes are too cumbersome. Either way will unfortunately require a lot of manual labor.

As a general trend, moving more stuff that needs configuration and logging (i.e., code that is in fact impure but masquerading as a top-level function) into methods would be a good trend.

**Comment by sampsyo** [comment]
Wonderful! I have merged the previous PR (`logging` branch) now, so feel free to open a new one for this stage (`logging-next`).

I want to make one minor suggestion about the log-level adjustment: I think we might be able to keep things simpler (and more backwards-compatible) if we make the per-plugin loggers `INFO` level by default (rather than `WARNING`). Then we only need to adjust the level for the import stages—commands can use the default level. (Note that wrapping the import stages is less fiddly than wrapping the command functions, which are one more level of indirection away.)

We should also update the plugin-authoring docs to reflect the new `self._log`.

**Comment by brunal** [comment]
New problem I met: event listeners.
One would usually do
```
@MyPlugin.listen("eventname")
def listener(foo, bar):
    log.debug("I listened!")
    pass
```
However now the listener doesn't have access to `log` any more... The obvious solution would be
```
class MyPlugin(BeetsPlugin)
    # ...
    @MyPlugin.listen("eventname")
    def listener(self, foo, bar):
        self._log.debug("I listened!")
        pass
````
But this is *not* possible: `MyPlugin` is not yet defined when we're doing `@MyPlugin.listen(...`.

The only decent solution I found is the following one:
````
class MyPlugin(BeetsPlugin)
    def __init__(self, *args):
        # ...
        self.register_listener("eventname", self.listener)

    def listener(self, foo, bar):
        self._log.debug("I listened!")
        pass
````
Is that ok with you? It *might* cause problems if we ever instantiate a given plugin twice.

Otherwise it's going fine. `INFO` is now the default level and `commands` is untouched. I'm moving big chunks of code around so they know the plugin instance though I don't feel like being known as the author of those lines by git... I'll update the docs as the last step.

**Comment by sampsyo** [comment]
Yep, this is indeed an issue, but I agree with your approach. (N.B. a few plugins even do this already: chroma, discogs, fetchart, freedesktop, ihate, and zero, according to a quick grep.) I actually regret doing the decorator thing in the first place—the upside was that the code looked clean for small examples, but the downside is that it introduced class-level state. This has turned out to be a headache, especially for testing. We should have used object-level state and left the class itself alone.

Methods are the way to go. I've been toying with a few metaclass tricks to make decorators work even for methods, but we can cross that bridge later. `register_listener` in `__init__` is the right way for now.

**Comment by brunal** [comment]
Is this issue still relevant? I feel like closing it now. It if can be closed then I believe #327 can be closed as well.

**Comment by sampsyo** [comment]
Yep. We still haven't done anything to unify "normal" logging and the importer log, but that's a somewhat different issue.
